src/System.Net.Utilities/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -2,33 +2,268 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System.ComponentModel;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
 using System.Threading;
+using System.Diagnostics;
+using System.Text;
+using System.IO;
 
 namespace System.Net.NetworkInformation
 {
     public partial class Ping
     {
-        private void InternalSendAsync(IPAddress address, byte[] buffer, int timeout, PingOptions options)
+        private const int IcmpHeaderLengthInBytes = 8;
+        private const int IpHeaderLengthInBytes = 20;
+
+        private static readonly bool s_canUseRawSockets = CheckRawSocketPermissions();
+
+        // Ubuntu has ping under /bin, OSX under /sbin.
+        private static readonly string[] s_binFolders = { "/bin", "/sbin" };
+        private static readonly string s_ipv4PingFile = "ping";
+        private static readonly string s_ipv6PingFile = "ping6";
+        private static readonly string s_discoveredPing4UtilityPath = GetPingUtilityPath(true);
+        private static readonly string s_discoveredPing6UtilityPath = GetPingUtilityPath(false);
+
+        private static bool CheckRawSocketPermissions()
+        {
+            try
+            {
+                Socket s = new Socket(AddressFamily.InterNetwork, SocketType.Raw, ProtocolType.Icmp);
+                return true;
+            }
+            catch
+            {
+                return false;
+            }
+        }
+
+        private static string GetPingUtilityPath(bool ipv4)
+        {
+            string fileName = ipv4 ? s_ipv4PingFile : s_ipv6PingFile;
+            foreach (string folder in s_binFolders)
+            {
+                string path = Path.Combine(folder, fileName);
+                if (File.Exists(path))
+                {
+                    return path;
+                }
+            }
+
+            return null;
+        }
+
+        private unsafe void InternalSendAsync(IPAddress address, byte[] buffer, int timeout, PingOptions options)
         {
             AsyncOperation asyncOp = _asyncOp;
             SendOrPostCallback callback = _onPingCompletedDelegate;
 
-            // TODO: Implement this (#2487)
+            PingReply pr = null;
+            Exception pingException = null;
+
+            try
+            {
+                if (s_canUseRawSockets)
+                {
+                    pr = SendIcmpPingOverRawSocket(address, buffer, timeout, options);
+                }
+                else
+                {
+                    pr = SendWithPingUtility(address, buffer, timeout, options);
+                }
+            }
+            catch (Exception e)
+            {
+                pingException = e;
+            }
 
-            PingException pe = new PingException(SR.net_ping, new PlatformNotSupportedException());
+            // At this point, either PR has a real PingReply in it, or pingException has an Exception in it.
```

Comment:
> Doesn't that mean this gets called on a background thread?

If the call to InternalSendAsync were wrapped in QueueUserWorkItem, then yes.  But it's not.

The simplest solution is to just use Task.Run or QueueUserWorkItem to surround the body of InternalSendAsync.  That's not the most efficient thing, though, and would be problematic if someone were, for example, writing a tool that initiated thousands of pings concurrently against all of the machines in their infrastructure, in which case each ping would effectively require its own thread.  The better solution would be to use the async functionality inherent to the underlying mechanism you're building on, e.g. Socket's async functionality rather than sync functionality.